File: core/obligations.py
"""Ageing an obligation: when is it due, and is it late.

This is the engine behind the claim the whole product rests on. The app already
knew what a certificate *said*; §4.3 is about knowing a certificate was **due and
never arrived**. For an FM contractor that inversion is the value — the risk is
the missing test, not the failed one.

Pure functions, no IO, no database. `today` is always a parameter and never
`date.today()`, so every result is reproducible and testable, and a report
rendered for an audit can be re-rendered identically later.

THREE KINDS OF OBLIGATION, which migration 023's obligations_cadence_check
enforces as mutually exclusive:

  periodic              a cadence is set; due again every N months or days
  event_triggered       no cadence; becomes due when a named event occurs
  self_declared_review  neither; the guideline states a duty but no frequency

The third exists because the guidelines demanded it. Loading the corpus found 41
obligations stating a duty with no frequency at all — "the grease trap must be
cleaned", with no interval. They cannot be aged from the document alone, and
pretending otherwise would either invent a deadline nobody agreed to or leave a
row that silently never comes due.

WHAT NEVER HAPPENS HERE
-----------------------
**No status is ever inferred to be `compliant` from missing data.** A periodic
obligation with no `next_due_on` has never been scheduled; that is a gap in
configuration, not a clean record, and it is reported as needing attention. The
same reasoning that made `status` NOT NULL with no default in 023, and that makes
`resolve_limits` return None rather than a default in §7.4: in a registry whose
job is noticing absence, the dangerous answer is the reassuring one.
"""
from __future__ import annotations

import calendar
from datetime import date, timedelta
from typing import Any, Mapping, NamedTuple, Optional

COMPLIANT = "compliant"
DUE_SOON = "due_soon"
OVERDUE = "overdue"
SUSPENDED = "suspended"
STATUSES = (COMPLIANT, DUE_SOON, OVERDUE, SUSPENDED)

PERIODIC = "periodic"
EVENT_TRIGGERED = "event_triggered"
SELF_DECLARED_REVIEW = "self_declared_review"

# How much warning "due soon" gives, by default. Capped against the cadence
# itself: a 30-day horizon on a 7-day inspection would leave it permanently
# due_soon, which is the same as no warning at all. A quarter of the interval is
# the compromise — enough notice to book a laboratory, short enough that the
# state still means something on a short cycle.
DEFAULT_DUE_SOON_DAYS = 30
DUE_SOON_FRACTION = 0.25


class ObligationStatus(NamedTuple):
    status: str
    kind: str
    reason: str
    days_until_due: Optional[int]   # negative once past due; None when unscheduled
    needs_attention: bool           # a data problem, distinct from being late


def add_months(start: date, months: int) -> date:
    """Add whole months, clamping to the end of a short month.

    31 January + 1 month is 28 February (or the 29th in a leap year), not
    3 March. A sampling obligation set on the 31st must not silently drift
    forward a few days every short month until it has skipped a period.
    """
    if months == 0:
        return start
    total = start.month - 1 + months
    year = start.year + total // 12
    month = total % 12 + 1
    day = min(start.day, calendar.monthrange(year, month)[1])
    return date(year, month, day)


def cadence_kind(obligation: Mapping[str, Any]) -> str:
    """Which of the three kinds this is, per 023's cadence CHECK."""
    if obligation.get("cadence_months") or obligation.get("cadence_days"):
        return PERIODIC
    if (obligation.get("trigger_event") or "").strip():
        return EVENT_TRIGGERED
    if obligation.get("self_declared_review"):
        return SELF_DECLARED_REVIEW
    # The CHECK makes this unreachable from the database, but this function is
    # also given un-persisted rows during loading.
    return SELF_DECLARED_REVIEW


def cadence_in_days(obligation: Mapping[str, Any]) -> Optional[int]:
    """Approximate interval length, for sizing the due-soon window only.

    Deliberately approximate — 30 days to a month. It is never used to compute a
    due DATE, which always goes through add_months so month lengths are exact.
    """
    if obligation.get("cadence_days"):
        return int(obligation["cadence_days"])
    if obligation.get("cadence_months"):
        return int(obligation["cadence_months"]) * 30
    return None


def next_due_after(satisfied_on: date, obligation: Mapping[str, Any]) -> Optional[date]:
    """When a periodic obligation falls due again, having just been satisfied.

    Measured from the date it was SATISFIED, not from the date it was previously
    due. Those differ whenever evidence arrives late, and measuring from the due
    date would stack a second deadline on top of a breach the client is already
    dealing with. Measuring from satisfaction is also how a laboratory schedule
    actually works.
    """
    if obligation.get("cadence_months"):
        return add_months(satisfied_on, int(obligation["cadence_months"]))
    if obligation.get("cadence_days"):
        return satisfied_on + timedelta(days=int(obligation["cadence_days"]))
    return None


def due_soon_window(obligation: Mapping[str, Any],
                    default_days: int = DEFAULT_DUE_SOON_DAYS) -> int:
    """How many days before the due date the obligation starts warning."""
    interval = cadence_in_days(obligation)
    if not interval:
        return default_days
    return max(1, min(default_days, int(interval * DUE_SOON_FRACTION)))


def evaluate(obligation: Mapping[str, Any], today: date,
             default_days: int = DEFAULT_DUE_SOON_DAYS) -> ObligationStatus:
    """Age one obligation as at `today`."""
    kind = cadence_kind(obligation)

    # A suspended obligation is not monitored and must not be aged. §7.5: a
    # client who un-ticks a module stops being monitored, which is commercially
    # correct — but it must be visibly suspended, never quietly compliant.
    if obligation.get("status") == SUSPENDED:
        return ObligationStatus(
            SUSPENDED, kind,
            "entitlement inactive — not monitored, and history retained",
            None, needs_attention=False)

    grace = int(obligation.get("grace_days") or 0)
    due = obligation.get("next_due_on")
    if isinstance(due, str):
        due = date.fromisoformat(due)

    if due is None:
        if kind == EVENT_TRIGGERED:
            # Nothing is owed until the event happens. This is the one case where
            # "no due date" genuinely means "nothing outstanding".
            return ObligationStatus(
                COMPLIANT, kind,
                f"awaiting trigger: {obligation.get('trigger_event')}",
                None, needs_attention=False)

        if kind == SELF_DECLARED_REVIEW:
            # The guideline states the duty but not the frequency, so nobody has
            # yet decided one. Not late — but not tracked either, and it must not
            # read as a clean record.
            return ObligationStatus(
                COMPLIANT, kind,
                "the guideline sets no frequency — a cadence must be agreed with "
                "the client before this can be tracked",
                None, needs_attention=True)

        # Periodic with no due date has never been scheduled. Reported as
        # overdue on purpose: in a registry whose job is noticing absence, an
        # unscheduled duty must demand attention rather than look satisfied.
        return ObligationStatus(
            OVERDUE, kind,
            "periodic obligation with no next_due_on — never scheduled",
            None, needs_attention=True)

    days = (due - today).days

    if today > due + timedelta(days=grace):
        late = (today - due).days
        return ObligationStatus(
            OVERDUE, kind,
            f"due {due.isoformat()}, {late} day(s) ago"
            + (f", grace of {grace} day(s) exhausted" if grace else ""),
            days, needs_attention=False)

    if days < 0:
        return ObligationStatus(
            DUE_SOON, kind,
            f"due {due.isoformat()}, within {grace}-day grace period",
            days, needs_attention=False)

    if days <= due_soon_window(obligation, default_days):
        return ObligationStatus(
            DUE_SOON, kind, f"due in {days} day(s), on {due.isoformat()}",
            days, needs_attention=False)

    return ObligationStatus(
        COMPLIANT, kind, f"next due {due.isoformat()}", days, needs_attention=False)


def satisfy(obligation: Mapping[str, Any], satisfied_on: date,
            evidence_id: Optional[str] = None,
            evidence_kind: Optional[str] = None) -> dict:
    """The column changes recording that evidence arrived.

    Returns a patch rather than mutating, so the caller decides whether to write
    it and the function stays pure. A periodic obligation gets its next due date;
    an event-triggered one goes back to awaiting its next trigger.

    `evidence_id` and `evidence_kind` are one fact and must be given together.
    023's obligations_satisfied_pair_check rejects half of it, and rightly: a kind
    with no id names a category of evidence nobody can produce, and an id with no
    kind cannot be resolved to a table, since §4.3 makes the reference polymorphic
    across lab_samples, certificates, inspections and risk_assessments with no
    foreign key to disambiguate it. Caught by the database on the first real run
    of this function.
    """
    if bool(evidence_id) != bool(evidence_kind):
        raise ValueError(
            "evidence_id and evidence_kind must be given together or not at all — "
            "023's obligations_satisfied_pair_check rejects a half-recorded "
            f"reference (got id={evidence_id!r}, kind={evidence_kind!r})")

    patch: dict[str, Any] = {
        "last_satisfied_at": satisfied_on.isoformat(),
        "last_satisfied_by": evidence_id or None,
        "last_satisfied_kind": evidence_kind or None,
    }
    nxt = next_due_after(satisfied_on, obligation)
    patch["next_due_on"] = nxt.isoformat() if nxt else None
    return patch


def summarise(obligations, today: date,
              default_days: int = DEFAULT_DUE_SOON_DAYS) -> dict:
    """Counts for a dashboard.

    `needs_attention` is counted separately and NOT folded into overdue. They are
    different conversations: one is "you are late", the other is "we cannot tell
    whether you are late", and merging them would let a configuration gap hide
    inside a compliance figure.
    """
    counts = {s: 0 for s in STATUSES}
    attention = 0
    for ob in obligations:
        result = evaluate(ob, today, default_days)
        counts[result.status] += 1
        if result.needs_attention:
            attention += 1
    counts["needs_attention"] = attention
    counts["total"] = sum(counts[s] for s in STATUSES)
    return counts

File: frontend/src/lib/obligations.ts
// Obligation registry — types and PRESENTATION ONLY.
//
// THE RULE THIS FILE EXISTS TO KEEP: nothing here decides whether a duty is
// late. `core/obligations.py:evaluate` does that, once, on the server, with
// `today` as a parameter so an audit can reproduce it. The API hands every row
// its `status`, `kind`, `reason`, `days_until_due` and `needs_attention`; this
// module turns those into a colour, a word and a sort order and does nothing
// else. There is no comparison against a limit or a date anywhere in this file,
// and there must never be one — `frontend/VERDICT_DIVERGENCE.md` documents what
// happens when the same verdict is implemented twice, and this repo is already
// at eight implementations (three of them in this frontend).
//
// The second rule, inherited from the endpoint: `needs_attention` is NOT
// overdue, is not folded into it, and is rendered separately everywhere.
// "You are late" and "we cannot tell whether you are late" are different
// sentences with different remedies.
//
// The third, inherited from lib/status.ts: green is reachable only on an
// explicit `compliant`. Anything unrecognised — including an absent status —
// renders slate and says so. A missing status must never look clean.

import type { BadgeTone } from '../components/ui/StatusBadge'

/** The four statuses `core/obligations.py` can return (023's status CHECK). */
export type ObligationStatusKey = 'compliant' | 'due_soon' | 'overdue' | 'suspended'

/** The three cadence kinds 023's obligations_cadence_check makes exclusive. */
export type ObligationKindKey = 'periodic' | 'event_triggered' | 'self_declared_review'

/** One row of GET /api/obligations. Verdict fields are marked; they are read,
 *  never recomputed. `status` and `kind` are typed as plain strings on purpose —
 *  an unrecognised value from a future server must be renderable, not a crash. */
export interface Obligation {
  id: string
  site_id: string | null
  /** Joined server-side from `sites`; null on an organisation-wide duty. */
  site_name: string | null
  entitlement_id: string | null
  label: string
  obligation_type: string | null
  notes: string | null
  next_due_on: string | null
  grace_days: number | null
  cadence_months: number | null
  cadence_days: number | null
  trigger_event: string | null
  self_declared_review: boolean | null
  last_satisfied_at: string | null
  /** No display name is returned with this — see the note in Obligations.tsx. */
  responsible_user_id: string | null

  // ── Computed by core.obligations.evaluate. Rendered verbatim. ──
  status: string
  /** The `status` COLUMN. Differs from `status` only when the ageing sweep is
   *  stale, which is a finding worth surfacing rather than hiding. */
  stored_status: string | null
  kind: string
  reason: string
  days_until_due: number | null
  needs_attention: boolean
}

/** The block `core.obligations.summarise` returns. */
export interface ObligationCounts {
  compliant: number
  due_soon: number
  overdue: number
  suspended: number
  /** Counted SEPARATELY and never inside `overdue`. */
  needs_attention: number
  total: number
}

export interface SiteCounts extends ObligationCounts {
  site_id: string | null
  /** "Organisation-wide" for site-less duties — labelled by the API, not dropped. */
  site_name: string | null
}

export interface ObligationsResponse {
  as_of: string
  site_id: string | null
  obligations: Obligation[]
  summary: ObligationCounts
}

export interface ObligationsSummaryResponse {
  as_of: string
  by_site: SiteCounts[]
  totals: ObligationCounts
}

export interface StatusPresentation {
  tone: BadgeTone
  label: string
  /** Section order: overdue first, and an unrecognised status is never filed
   *  below `compliant` where it would read as a quiet pass. */
  order: number
  /** Plain-language gloss shown once per section, not per row. */
  blurb: string
}

/**
 * A colour and a word for a status the server already decided.
 *
 * Written as an exhaustive switch with an explicit default rather than a lookup
 * table, so an unknown status cannot land on a green entry by accident and
 * cannot throw. `green` is returned for exactly one input.
 */
export function presentStatus(status: string | null | undefined): StatusPresentation {
  switch (status) {
    case 'overdue':
      return {
        tone: 'red', label: 'Overdue', order: 0,
        blurb: 'Past the due date (and any grace period). Evidence is owed now.',
      }
    case 'due_soon':
      return {
        tone: 'amber', label: 'Due soon', order: 1,
        blurb: 'Inside the warning window, or inside the grace period after the due date.',
      }
    case 'suspended':
      return {
        tone: 'slate', label: 'Suspended', order: 3,
        blurb: 'The entitlement is inactive, so this is not monitored. Deliberately not '
          + '"compliant": a commercial decision is not a clean compliance record. History is retained.',
      }
    case 'compliant':
      return {
        tone: 'green', label: 'Compliant', order: 4,
        blurb: 'Scheduled and not yet inside the warning window, or awaiting its trigger.',
      }
    default:
      return {
        tone: 'slate',
        label: status ? `Unrecognised status: ${status}` : 'No status returned',
        order: 2,
        blurb: 'The server returned a status this build does not recognise. It is shown '
          + 'as unjudged — never as a pass — until the vocabulary is reconciled.',
      }
  }
}

export interface KindPresentation {
  label: string
  description: string
}

/** How the duty comes due. Wording only — the server chose the kind. */
export function presentKind(kind: string | null | undefined): KindPresentation {
  switch (kind) {
    case 'periodic':
      return { label: 'Periodic', description: 'Falls due again on a fixed cadence.' }
    case 'event_triggered':
      return { label: 'Event-triggered', description: 'Becomes due when a named event occurs.' }
    case 'self_declared_review':
      return {
        label: 'Self-declared review',
        description: 'The guideline states the duty but no frequency.',
      }
    default:
      return { label: kind || 'Unknown', description: 'Cadence kind not recognised by this build.' }
  }
}

export interface AttentionPresentation {
  /** Short badge text. Never the word "overdue". */
  label: string
  /** What the user has to DO about it. */
  remedy: string
}

/**
 * The separate "we cannot tell" signal.
 *
 * Returns null when the row does not carry it — the flag comes from the server,
 * this only chooses the wording, keyed off the server's `kind`. A duty whose
 * guideline states no frequency is an unanswered question, not a breach, and
 * the remedy is a conversation, so the remedy is what is shown.
 */
export function presentAttention(obligation: Obligation): AttentionPresentation | null {
  if (!obligation.needs_attention) return null
  switch (obligation.kind) {
    case 'self_declared_review':
      return {
        label: 'Cadence to agree',
        remedy: 'The guideline sets no frequency for this duty, so it cannot be aged. '
          + 'This is not a breach. Agree a cadence with the client and record it, and '
          + 'this duty starts being tracked.',
      }
    case 'periodic':
      return {
        label: 'Never scheduled',
        remedy: 'This duty has a cadence but no due date has ever been set, so its ageing '
          + 'cannot be trusted. Give it a first due date.',
      }
    default:
      return { label: 'Needs attention', remedy: obligation.reason }
  }
}

/** `days_until_due` as words. Formatting of a number the server computed. */
export function formatDaysUntilDue(days: number | null | undefined): string {
  if (days === null || days === undefined) return 'No due date'
  if (days === 0) return 'Due today'
  if (days < 0) return `${Math.abs(days)} day${Math.abs(days) === 1 ? '' : 's'} late`
  return `in ${days} day${days === 1 ? '' : 's'}`
}

/** Colour for the days-until-due cell — taken from the status the server gave,
 *  never from the sign of the number, so the two can never disagree. */
export function daysTone(obligation: Obligation): BadgeTone {
  return presentStatus(obligation.status).tone
}

/**
 * Sort inside a section: soonest first, unscheduled last.
 *
 * Ordering by a number the server supplied. It decides nothing — every row in a
 * section already carries the same server-issued status.
 */
export function byUrgency(a: Obligation, b: Obligation): number {
  const av = a.days_until_due
  const bv = b.days_until_due
  if (av === null || av === undefined) return bv === null || bv === undefined ? 0 : 1
  if (bv === null || bv === undefined) return -1
  return av - bv
}

/** Group rows into sections in `presentStatus.order`, ready to render. */
export function groupByStatus(
  obligations: Obligation[],
): { status: string; presentation: StatusPresentation; rows: Obligation[] }[] {
  const buckets = new Map<string, Obligation[]>()
  for (const o of obligations) {
    const key = o.status ?? ''
    const bucket = buckets.get(key)
    if (bucket) bucket.push(o)
    else buckets.set(key, [o])
  }
  return [...buckets.entries()]
    .map(([status, rows]) => ({
      status,
      presentation: presentStatus(status),
      rows: [...rows].sort(byUrgency),
    }))
    .sort((a, b) => a.presentation.order - b.presentation.order)
}

// ── Module catalogue (GET /api/modules) ──────────────────────────────────────

export interface GuidelineModule {
  id: string
  key: string | null
  label: string | null
  category: string | null
  standard_id: string | null
  /** §7.12 — what a report may CLAIM for this module. Only 'compliance' may
   *  produce a verdict at all. */
  module_kind: string | null
  obligation_type: string | null
  status: string | null
  provenance: string | null
  notes: string | null
  entitled: boolean
  entitlement_id: string | null
  active_from: string | null
  active_until: string | null
  /** 023 refuses status='available' unless provenance='verified'. */
  sellable: boolean
  /** Why not, in words, straight from the server. Null when it is sellable. */
  not_sellable_reason: string | null
  /** Present only for roles holding billing.read. */
  list_price_monthly?: number | null
  currency?: string | null
}

export interface ModulesResponse {
  modules: GuidelineModule[]
  entitled_count: number
  sellable_count: number
}

/** What ticking WOULD create — `core.entitlements.plan_summary`, verbatim. */
export interface EntitlementPlan {
  module: { id: string; key: string | null; label: string | null; module_kind: string | null }
  sites: number
  obligations: {
    total: number
    by_kind: Record<string, number>
    needs_cadence_agreed: number
    due_immediately: number
    awaiting_trigger: number
  }
  warning: string
}

export interface EntitlementPlanResponse {
  created: boolean
  plan: EntitlementPlan
  message?: string
  entitlement?: { id: string; active_from: string; active_until: string | null }
  obligations_created?: number
}

/** One duty that stops being tracked when a module is un-ticked (§7.5). */
export interface NoLongerMonitored {
  id: string
  label: string | null
  site_id: string | null
  next_due_on: string | null
}

export interface DeactivateResponse {
  deactivated: boolean
  obligations_suspended: number
  obligations_deleted: number
  no_longer_monitored: NoLongerMonitored[]
  message: string
}

export interface ModuleKindPresentation {
  label: string
  tone: BadgeTone
  description: string
}

/** §7.12 module kind — what the module is allowed to claim. */
export function presentModuleKind(kind: string | null | undefined): ModuleKindPresentation {
  switch (kind) {
    case 'compliance':
      return {
        label: 'Compliance', tone: 'green',
        description: 'Sets testable limits, so a report may state a compliance verdict.',
      }
    case 'procedural':
      return {
        label: 'Procedural', tone: 'blue',
        description: 'States duties and practice, not limits. A report may evidence the duty '
          + 'was done — it may not claim a compliance verdict.',
      }
    case 'advisory':
      return {
        label: 'Advisory', tone: 'blue',
        description: 'Guidance only. Sellable, but not the same product as a compliance module.',
      }
    case 'unusable':
      return {
        label: 'Unusable', tone: 'red',
        description: 'The guideline contradicts itself — there is nothing a report could '
          + 'truthfully say against it.',
      }
    default:
      return {
        label: kind || 'Unknown', tone: 'slate',
        description: 'Module kind not recognised by this build.',
      }
  }
}

/** Catalogue lifecycle status. Not a compliance verdict — never green here
 *  except for a module actually on sale. */
export function presentModuleStatus(status: string | null | undefined): { label: string; tone: BadgeTone } {
  switch (status) {
    case 'available': return { label: 'Available', tone: 'green' }
    case 'coming_soon': return { label: 'Coming soon', tone: 'blue' }
    case 'retired': return { label: 'Retired', tone: 'slate' }
    default: return { label: status || 'Unknown', tone: 'slate' }
  }
}

/** Provenance of the module's content against the published DM document. */
export function presentProvenance(provenance: string | null | undefined): { label: string; tone: BadgeTone } {
  switch (provenance) {
    case 'verified': return { label: 'Verified', tone: 'green' }
    case 'unverified': return { label: 'Unverified', tone: 'amber' }
    case 'extracted': return { label: 'Extracted, unread', tone: 'amber' }
    default: return { label: provenance || 'Unknown', tone: 'slate' }
  }
}
